add in-game console and hs compiler logic

Berthalamew opened the pull request:
- adds logic to run commands at startup using init.txt

<img width="2547" height="1329" alt="image" src="https://github.com/user-attachments/assets/33cc3193-468b-4aa1-9502-79e5cd711afd" />

Berthalamew committed:
add rectangle2d helper functions

Berthalamew committed:
remove math module

- move macros to cseries

Berthalamew committed:
add timeout code to discord interface initialization

- make sure we don't get caught in a deadlock if we're stuck initializing

Berthalamew committed:
add ai verification functions and other ai logic for terminal

Berthalamew committed:
add cseries error logic

- rename shell_update to shell_idle

Berthalamew committed:
math changes

- rename dequantize_unit_vector to dequantize_unit_vector3d
- add real_to_long_round
- add event in dequantize_unit_vector3d

Berthalamew committed:
add tmpl_IN_RANGE function

Berthalamew committed:
misc additions

- add debug menu functions stubs
- add editor stubs
- add string constants for game enum names
- add player_mapping_first_active_output_user
- add debug keys stubs
- add stub for tag_group_dependencies_register_dependency
- formatting changes

Berthalamew committed:
tag changes

- add string id functions

Berthalamew committed:
add overload for rasterizer_draw_string

Berthalamew committed:
add edit_text functionality

Berthalamew committed:
cseries additions and other misc changes

- rename observer_update_internal to observer_apply_camera_effect
- add handle_fatal_error function
- adjust assertion macros
- remove and operation on DATUM_INDEX_TO_IDENTIFIER
- add assert_return macro
- rename debugging boolean globals
- add address/pointer helper functions
- add device_group_get_from_scenario_index
- update first person weapon structures and names
- add hud definitions and constants
- add debug functions in user interface
- fix incorrect macro usage in crash_window_create_text
- make Xliverendering logic use critical sections instead of mutex and lock guard

Berthalamew committed:
reimplement some data functions and add asserts

Berthalamew committed:
add halo script compiler and console logic

nukeulater commented on xlive/Blam/Engine/math/integer_math.h:
```diff
@@ -39,6 +39,35 @@ inline int16 rectangle2d_height(const rectangle2d* rect)
+inline rectangle2d* offset_rectangle2d(rectangle2d* rectangle, int16 dx, int16 dy)
+{
+	rectangle->left += dx;
+	rectangle->left += dx;
```
Probably unintended duplicate line

Berthalamew replied:
Yup, thanks for catching

Berthalamew replied:
fixed

nukeulater commented on xlive/Blam/Engine/math/real_quantization.cpp:
```diff
@@ -68,13 +71,13 @@ int32 quantize_unit_vector(const real_vector3d* u)
 {
-	real32 dequantized_f1 = dequantize_real((uint8)(quantized_unit_vector >>  3), -1.0f, 1.0f, 8, true);
-	real32 dequantized_f2 = dequantize_real((uint8)(quantized_unit_vector >> 11), -1.0f, 1.0f, 8, true);
+	real32 dequantized_f1 = dequantize_real((uint8)((quantized_unit_vector >>  3) & BYTE_MAX), -1.0f, 1.0f, 8, true);
```
Explain

Berthalamew replied:
Removes the upper bits of the vector before casting to a uint8, caused errors when using the compiler flag to break when a value was truncated due to type size. Should probably replace this with MAX.

Berthalamew replied:
MAX would break this, going to leave it so we're explicit that we want those 8 bits and we don't want to take anything above that into account.

Berthalamew replied:
added comment in code to explain this

nukeulater commented on xlive/H2MOD/Modules/MapManager/MapManager.h:
```diff
@@ -21,7 +20,7 @@ class MapDownloadQuery {
 	void StopDownload();
 
-	std::wstring m_clientMapFilenameWide;
+	wchar_t m_clientMapFilenameWide[32];
```
Do you guarantee a max of 32 character file names?

nukeulater replied:
Also 32 instead of a defined constant...

Berthalamew replied:
Got the number wrong here, it was 48 and I yea I need to make a constant for it.

Berthalamew replied:
nevermind I was correct initially, they just have the hash appended to it as well, I'll need to adjust the game options struct too.

Berthalamew replied:
updated

nukeulater replied:
This is the filename on the disk iirc, not the internal one, which could have many more characters.

nukeulater commented on xlive/Blam/Engine/physics/character_physics_mode_melee.cpp:
```diff
@@ -180,8 +179,7 @@ void __thiscall c_character_physics_mode_melee_datum::update_internal
 				m_has_target = true;
-				m_maximum_counter = blam_ticks_real_to_integer((real32)(m_weapon_is_sword ? 7 : 1) + lunge_tick_count_adjusted);
-				// m_maximum_counter = (real32)(m_weapon_is_sword ? 7 : 1) + lunge_tick_count_adjusted;
+				m_maximum_counter = (int32)((m_weapon_is_sword ? 7.f : 1.f) + lunge_tick_count_adjusted);
```
Heh.. 

Did you even check what blam_ticks_real_to_int() does?

Berthalamew replied:
Yes, however the actual assembly and instructions used differ between the builds of the game.
This is probably just because they didn't append `.f` to the numbers in the ternary statement. Since they add the result to a float and then brought down to an int it needs to be truncated.

nukeulater replied:
It should be ceil truncated, not floor truncated, which is the default behavior.

nukeulater commented on xlive/Blam/Engine/hs/hs_runtime.cpp:
```diff
@@ -0,0 +1,1311 @@
+
+static int32 hs_short_to_real(int32 s)
+{
+	const real32 v = (real32)(int16)s;
```
These don't really look right to me

Berthalamew replied:
This is correct

nukeulater commented on xlive/Blam/Engine/hs/hs_runtime.cpp:
```diff
@@ -0,0 +1,1311 @@
+
+static int32 hs_real_to_short(int32 r)
+{
+	return (int16)(int32)*(real32*)&r;
```
Forcefully reinterpreting a float to a 32 bit value then downcasting it to 16 bit, while the param is not even a float doesn't sound like the intended behavior, same with the other functions...

I'd double check again if the implementation is valid

twist84 replied:
Unfortunately with the way Bungie did callbacks for hs casting this is how it is

Berthalamew replied:
What twister mentioned is correct

nukeulater commented on xlive/Blam/Engine/cseries/cseries.h:
```diff
@@ -124,41 +137,32 @@ static_assert (sizeof(STRUCT) == (_SIZE), "Invalid size for struct ("#STRUCT") e
-ASSERT_TRIGGER_EXCEPTION(IS_EXCEPTION);							\
+#define DISPLAY_ASSERT_EXCEPTION(STATEMENT, FILE, LINE, IS_EXCEPTION)	\
+display_assert(STATEMENT, FILE, LINE, IS_EXCEPTION);			\
+if (is_debugger_present())												\
```
No do while(0) guards

Berthalamew replied:
changed

Berthalamew committed:
custom map changes

- add constant for custom name length
- rename s_custom_map_id and fields to s_secure_map_id
- update game options struct and enums

Berthalamew committed:
update quantization code

- add enum for face types
- add asserts

Berthalamew committed:
misc changes

- add some comments in haloscript code
- remove incorrect render window enum entries
- remove CONSOLE_ENABLED usage (unused)
- fix duplicate line in offset_rectangle2d
- rename c_static_array_tracked to c_static_stack
- add render_frame_present

Berthalamew committed:
fix bug when finding hs globals

- would use the external global count as a limit instead of the hs_globals tagblock count

Berthalamew committed:
misc changes

- make assert use assume compiler hints when in release mode
- add macros for getting specific object types
- add asserts in object related functions
- remove duplicate function to get the screen tag index
- add missing hs_return calls back into interface debug functions
- update account create error code to use a switch case instead of multiple if else if statements

nukeulater commented on xlive/Blam/Engine/math/real_math.cpp:
```diff
@@ -102,6 +102,11 @@ void __cdecl real_math_reset_precision(void)
 
+int32 real_to_long_round(real32 v)
+{
+	return (int32)v;
```
The standard rounding mode is rounding down, i.e floor or discarding the decimal places. 

We need to round nearest, which the previous code did (+-0.5), or using lrintf(), but only when the FPU is properly configured with the round control flag set to nearest. I hope I'll adjust some of the floating point math configuration soon.

Berthalamew committed:
add call to real_to_long_round in main_time_update

Berthalamew committed:
call real_to_long_round in quantize_real
